File: 213-HouseRobberII/HouseRobber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n=nums.size();
        if(n==0)    return 0;
        if(n==1)    return nums[0];
        
        int x1=0;
        int pre1=0;
        
        int x2=0;
        int pre2=0;
        
        int x;
        
        for(int i=0;i<n-1;i++){
            x=max(x1,pre1+nums[i]);
            pre1=x1;
            x1=x;
        }
        
        for(int i=1;i<n;i++){
            x=max(x2,pre2+nums[i]);
            pre2=x2;
            x2=x;
        }
        
        return max(x1,x2);
    }
};

File: 198-HouseRobber/houseRobber.java
public class Solution {
    public int rob(int[] nums) {
        if (nums.length == 0) return 0;
        //by induction
        int[] bestUpToN = new int[nums.length];
        bestUpToN[0] = nums[0];
        if ( nums.length > 1) {
            bestUpToN[1] = nums[0] < nums[1] ? nums[1] : nums[0];
        }
        for (int i = 2; i< nums.length; i++) {
            int temp = bestUpToN[i-2] + nums[i];//if broke ith house.
            //if don't broke ith house, best value is bestUpToN[i-1]
            //compare both value
            bestUpToN[i]= temp < bestUpToN[i-1] ? bestUpToN[i-1] : temp;
        }
        
        return bestUpToN[nums.length -1];
        
    }
}